### PR TITLE
JEP 9 - Boolean expressions

### PR DESCRIPTION
**Pull request by jamesls**
This jep adds improved boolean expressions and filters to JMESPath.

High level summary is it adds:

* And expression: `a && b`
* Not expressions: `!is_admin`
* Parens: `(a || b) && c`

In addition to top level expressions, you can also use them in filter expressions: `Users[? is_admin && !new_account]`

Will link jmespath.test updates and JEP 9 implementations to the tracking PR.

### CODE REVIEW

**Comment by jamesls** [comment]
Last call for feedback.  I think as far changes go, this is a pretty straightforward change that fleshes out the partial boolean support that JMESPath previously had.

If you want something more concrete to look at, here at the proposed compliance tests for this change: https://github.com/jmespath/jmespath.test/commit/f8f678ea9534178cc4ed43c85048b884a8b84c61

Note: I'll update the compliance tests to not use the outdated syntax of:


```
`foo`
```

File: docs/proposals/improved-filters.rst [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+* Or - ``||``
+* And - ``&&``
+* Unary not - ``!``
+
+So for example, ``a || b && c`` is parsed as ``a || (b && c)`` and
+not ``(a || b) && c``.
+
+The operator precedence list in the specification will now read:
+
+* Pipe - ``||``
```
Comment (mtdowling) [inline docs/proposals/improved-filters.rst]: Should be ``|``

File: docs/proposals/improved-filters.rst [hunk 2]
```diff
@@ -0,0 +1,345 @@
+================
+Improved Filters
+================
+
+:JEP: 9
+:Author: James Saryerwinnie
+:Status: proposed
+:Created: 07-July-2014
```
Comment (mtdowling) [inline docs/proposals/improved-filters.rst]: ?

**Comment by mtdowling** [comment]
Nice work. :shipit:

File: docs/proposals/improved-filters.rst (hunk 1 above)
Comment (jamesls) [inline docs/proposals/improved-filters.rst]: Fixed.

File: docs/proposals/improved-filters.rst (hunk 2 above)
Comment (jamesls) [inline docs/proposals/improved-filters.rst]: Yeah, it's been a while since this was first written.

**Comment by jamesls** [comment]
Ok, I'll mark this as accepted and merge.

**Pull request merged**